packages/frameworks-nextjs/src/client.tsx

```diff
@@ -0,0 +1,495 @@
+ * :::
+ */
+export function useSession<R extends boolean>(
+  options?: UseSessionOptions<R>
+): SessionContextValue<R> {
+  if (!SessionContext) {
+    throw new Error("React Context is unavailable in Server Components")
+  }
+
+  // @ts-expect-error Satisfy TS if branch on line below
+  const value: SessionContextValue<R> = React.useContext(SessionContext)
+  if (!value && process.env.NODE_ENV !== "production") {
+    throw new Error(
+      "[next-auth]: `useSession` must be wrapped in a <SessionProvider />"
+    )
+  }
+
+  const { required, onUnauthenticated } = options ?? {}
+
+  const requiredAndNotLoading = required && value.status === "unauthenticated"
+
+  React.useEffect(() => {
+    if (requiredAndNotLoading) {
+      const url = `/api/auth/signin?${new URLSearchParams({
+        error: "SessionRequired",
+        callbackUrl: window.location.href,
+      })}`
+      if (onUnauthenticated) onUnauthenticated()
+      else window.location.href = url
+    }
+  }, [requiredAndNotLoading, onUnauthenticated])
+
+  if (requiredAndNotLoading) {
+    return {
+      data: value.data,
+      update: value.update,
+      status: "loading",
+    }
+  }
+
+  return value
+}
+
+export interface GetSessionParams {
+  event?: "storage" | "timer" | "hidden" | string
+  triggerEvent?: boolean
+  broadcast?: boolean
+}
+
+export async function getSession(params?: GetSessionParams) {
+  const session = await fetchData<Session>(
+    "session",
+    __NEXTAUTH,
+    logger,
+    params
+  )
+  if (params?.broadcast ?? true) {
+    broadcast.postMessage({ event: "session", data: { trigger: "getSession" } })
+  }
+  return session
+}
+
+/**
+ * Returns the current Cross-Site Request Forgery Token (CSRF Token)
+ * required to make requests that changes state. (e.g. signing in or out, or updating the session).
+ *
+ * [CSRF Prevention: Double Submit Cookie](https://cheatsheetseries.owasp.org/cheatsheets/Cross-Site_Request_Forgery_Prevention_Cheat_Sheet.html#double-submit-cookie)
+ * @internal
+ */
+export async function getCsrfToken() {
+  const response = await fetchData<{ csrfToken: string }>(
+    "csrf",
+    __NEXTAUTH,
+    logger
+  )
+  return response?.csrfToken
+}
+
+type ProvidersType = Record<
+  LiteralUnion<BuiltInProviderType>,
+  ClientSafeProvider
+>
+
+/**
+ * Returns a client-safe configuration object of the currently
+ * available providers.
+ * @internal
+ */
+export async function getProviders() {
+  return fetchData<ProvidersType>("providers", __NEXTAUTH, logger)
+}
+
+/**
+ * Initiate a signin flow or send the user to the signin page listing all possible providers.
+ * Handles CSRF protection.
+ */
+export async function signIn<
+  P extends RedirectableProviderType | undefined = undefined
+>(
+  provider?: LiteralUnion<
+    P extends RedirectableProviderType
+      ? P | BuiltInProviderType
+      : BuiltInProviderType
+  >,
+  options?: SignInOptions,
```

Comment:
Can we use a type to make it obvious that, if the provider type is `email`, then `options` should include `email` ([v4 docs](https://next-auth.js.org/getting-started/client#starts-email-sign-in-flow-when-clicked))? Maybe something like:

```ts
options?: P extends "email"
  ? SignInOptions & { email: string }
  : SignInOptions,
```